account-integrations/safe/src/SafeZkEmailRecoveryPlugin.sol

```diff
@@ -0,0 +1,133 @@
+
+    mapping(address => ZkEmailRecoveryStorage) public zkEmailRecoveryStorage;
+
+    error INVALID_RECOVERY_HASH(
+        bytes32 recoveryHash,
+        bytes32 expectedRecoveryHash
+    );
+    error MODULE_NOT_ENABLED();
+    error INVALID_OWNER(address expectedOwner, address owner);
+    error INVALID_PROOF();
+
+    MockGroth16Verifier public immutable verifier;
+
+    constructor(address _verifier) {
+        verifier = MockGroth16Verifier(_verifier);
+        RECOVERY_HASH_DOMAIN = keccak256(
+            abi.encodePacked(
+                DOMAIN_NAME,
+                DOMAIN_VERSION,
+                block.chainid,
+                address(this)
+            )
+        );
+    }
+
+    /**
+     * @notice returns storage accociated with a safe address
+     * @param safe address to query storage with
+     */
+    function getZkEmailRecoveryStorage(
+        address safe
+    ) external view returns (ZkEmailRecoveryStorage memory) {
+        return zkEmailRecoveryStorage[safe];
+    }
+
+    /**
+     * @notice stores a recovery hash that can be used to recover a ecdsa plugin at a later stage.
+     * @dev this function assumes it is being called from a safe - see how msg.sender is interpreted.
+     * @param recoveryHash hash of domain, email and salt - keccak256(abi.encodePacked(RECOVERY_HASH_DOMAIN, email, salt))
+     * @param owner owner of the ecdsa plugin
+     * @param ecsdaPlugin safe ecsda plugin address that this function will be adding a recovery option for
+     */
+    function addRecoveryHash(
+        bytes32 recoveryHash,
+        address owner,
+        address ecsdaPlugin
+    ) external {
+        address safe = msg.sender;
+
+        bool moduleEnabled = ISafe(safe).isModuleEnabled(address(this));
+        if (!moduleEnabled) revert MODULE_NOT_ENABLED();
+
+        address expectedOwner = ISafeECDSAPlugin(ecsdaPlugin).getOwner(safe);
+        if (owner != expectedOwner) revert INVALID_OWNER(expectedOwner, owner);
+
+        zkEmailRecoveryStorage[safe] = ZkEmailRecoveryStorage(recoveryHash);
+    }
+
+    /**
+     * @notice recovers a safe ecdsa plugin using a zk email proof.
+     * @dev rotates the safe ecdsa plugin owner address to a new address. The email address hash is kept hidden until this function is called.
+     * This function is designed so it can be called from any account and account type.
+     * @param safe the safe that manages the safe ecdsa plugin being recovered
+     * @param ecdsaPlugin safe ecsda plugin address that this function will be rotating the owner address for
+     * @param newOwner the new owner address of the safe ecdsa plugin
+     * @param salt the salt used in the recovery hash
+     * @param email the email address hash used in the recovery hash
+     * @param a part of the proof
+     * @param b part of the proof
+     * @param c part of the proof
+     * @param publicSignals public inputs for the zkp
+     */
+    function recoverAccount(
```

Comment:
@jacque006 for the recovery functionality. I did a check on the recovery hash and then mocked the proof verification. It would be good to get your thoughts on if anything else is needed in this function. For example:

1. One thing I'm not sure on is when and how we query the dkim registry in the flow. 
2. The ZkEmailSafe verification appears to [verify the RSA signature](https://github.com/javiersuweijie/zkemail-safe/blob/13c5b3ca6c46cee8e1c0592ed4b072cb5b79ec9e/contracts/src/ZkEmailSafe.sol#L106C16-L106C16), but I remember reading elsewhere that was done inside the proof, so wasn't sure on whether to add that

Reply:
See comment below.